internal/pkg/vm/internal/vmcontext/vmcontex2.go

```diff
@@ -0,0 +1,543 @@
+	}
+
+	// Dragons: from address need to be normalized here
+
+	// 2. load actor from global state
+	fromActor, err := vm.state.GetActor(context.Background(), msg.From)
+	// Review: when are errors surfaced? this can lead to invalid state transitions
+	if fromActor == nil || err != nil {
+		// Execution error; sender does not exist at time of message execution.
+		return message.Failure(exitcode.ActorNotFound), gasTank.GasConsumed().Cost(msg.GasPrice)
+	}
+
+	// 3. make sure this is the right message order for fromActor
+	if msg.CallSeqNum != fromActor.CallSeqNum {
+		// Execution error; invalid seq number.
+		return message.Failure(exitcode.InvalidCallSeqNum), gasTank.GasConsumed().Cost(msg.GasPrice)
+	}
+
+	// 4. Check sender balance (gas + value being sent)
+	gasLimitCost := msg.GasLimit.Cost(msg.GasPrice)
+	totalCost := msg.Value.Add(gasLimitCost)
+	if fromActor.Balance.LessThan(totalCost) {
+		// Execution error; sender does not have sufficient funds to pay for the gas limit.
+		return message.Failure(exitcode.InsufficientFunds), gasTank.GasConsumed().Cost(msg.GasPrice)
+	}
+
+	// 5. Increment sender CallSeqNum
+	fromActor.IncrementSeqNum()
+
+	// 6. Deduct gas limit funds from sender first
+	// Note: this should always succeed, due to the sender balance check above
+	// Note: after this point, we nede to return this funds back before exiting
+	vm.transfer(msg.From, address.BurntFundsAddress, gasLimitCost)
+
+	// 7. checkpoint state
+	// Even if the message fails, the following accumulated changes will be applied:
+	// - CallSeqNumber increment
+	// - sender balance withheld
+	// - (if did not previously exist, the new auto-created target actor)
+
+	// 7.a. checkpoint store
+	// TODO: vm.store.Checkpoint() (issue: #3718)
+	// 7.b. checkpoint actor state
+	// TODO: vm.state.Checkpoint() (issue: #3718)
+
+	// send
+	// 1. build internal message
+	// 2. build invocation context
+	// 3. process the msg
+
+	// 1. build internal msg
+	imsg := internalMessage{
+		miner:  miner,
+		from:   msg.From,
+		to:     msg.To,
+		value:  msg.Value,
+		method: msg.Method,
+		params: msg.Params,
+	}
+
+	// 2. build invocation context
+	ctx := newInvocationContext(vm, imsg, fromActor, &gasTank)
+
+	// 3. process the msg (safely)
+	// Note: we use panics throughout the execution to signal early termination,
+	//       we need to trap them on this top level call
+	safeProcess := func() (out message.Receipt) {
+		// trap aborts and exitcodes
+		defer func() {
+			// TODO: discard "non-checkpointed" state changes (issue: #3718)
+
+			if r := recover(); r == nil {
+				switch r.(type) {
+				case runtime.AbortPanicError:
+					// Review: abort has a msg inside.. should we log it or what?
+					out = message.Failure(exitcode.MethodAbort).WithGas(gasTank.GasConsumed())
+					return
+				case exitcode.ExitCode:
+					out = message.Failure(r.(exitcode.ExitCode)).WithGas(gasTank.GasConsumed())
```

Comment:
I think these message.Failure call sites would be clearer if it just took two arguments, with gas value explicitly zero for the ones above where the miner pays.